jgrapht-core/src/main/java/org/jgrapht/alg/scoring/PageRank.java

```diff
@@ -0,0 +1,342 @@
+        }
+        return scores.get(v);
+    }
+
+    private void run(double dampingFactor, int maxIterations, double tolerance)
+    {
+        // initialization
+        Specifics specifics;
+        if (g instanceof DirectedGraph<?, ?>) {
+            specifics = new DirectedSpecifics((DirectedGraph<V, E>) g);
+        } else {
+            specifics = new UndirectedSpecifics(g);
+        }
+
+        int totalVertices = g.vertexSet().size();
+        boolean weighted = (g instanceof WeightedGraph<?, ?>);
+        Map<V, Double> weights;
+        if (weighted) {
+            weights = new HashMap<>(totalVertices);
+        } else {
+            weights = Collections.emptyMap();
+        }
+
+        double initScore = 1.0d / totalVertices;
+        for (V v : g.vertexSet()) {
+            scores.put(v, initScore);
+            if (weighted) {
+                double sum = 0;
+                for (E e : specifics.outgoingEdgesOf(v)) {
+                    sum += g.getEdgeWeight(e);
+                }
+                weights.put(v, sum);
+            }
+        }
+
+        // run PageRank
+        Map<V, Double> nextScores = new HashMap<>();
+        double maxChange = tolerance;
+
+        while (maxIterations > 0 && maxChange >= tolerance) {
+            // compute next iteration scores
+            double r = 0d;
+            for (V v : g.vertexSet()) {
+                if (specifics.outgoingEdgesOf(v).size() > 0) {
+                    r += (1d - dampingFactor) * scores.get(v);
+                } else {
+                    r += scores.get(v);
+                }
+            }
+            r /= totalVertices;
+
+            maxChange = 0d;
+            for (V v : g.vertexSet()) {
+                double contribution = 0d;
+
+                if (weighted) {
+                    for (E e : specifics.incomingEdgesOf(v)) {
+                        V w = Graphs.getOppositeVertex(g, e, v);
+                        contribution += dampingFactor * scores.get(w)
+                            * g.getEdgeWeight(e) / weights.get(w);
+                    }
+                } else {
+                    for (E e : specifics.incomingEdgesOf(v)) {
+                        V w = Graphs.getOppositeVertex(g, e, v);
+                        contribution += dampingFactor * scores.get(w)
+                            / specifics.outgoingEdgesOf(w).size();
+                    }
+                }
+
+                double vOldValue = scores.get(v);
+                double vNewValue = r + contribution;
+                maxChange = Math
+                    .max(maxChange, Math.abs(vNewValue - vOldValue));
+                nextScores.put(v, vNewValue);
+            }
+
+            // swap scores
+            Map<V, Double> tmp = scores;
+            scores = nextScores;
+            nextScores = tmp;
+
+            // progress
+            maxIterations--;
+        }
+
+    }
+
+    abstract class Specifics
```

Comment:
This pattern is starting to get copy/pasted throughout the codebase...we ought to factor it out to a single class reusable by all algorithms.